### ISSUE TITLE
Need cl_khr_gl_sharing

### ISSUE THREAD

**Issue by zhoub** [intel/compute-runtime#166]
Hi

I found that this driver doesn't support *cl_khr_gl_sharing*, is there any plan for this ? Thanks !

**Comment by PiotrRozenfeld** [comment]
At this point this extension is supported only on Windows. There are currently no plans to implement this extension on Linux.

**Comment by zhoub** [comment]
Hi

This is actually important as Windows platform. So many video applications require the OpenCL and OpenGL interop, and now our work is based on Intel Up board. Please consider this. Thanks !

**Comment by zhoub** [comment]
```
Number of platforms                               3
  Platform Name                                   Intel(R) OpenCL HD Graphics
  Platform Vendor                                 Intel(R) Corporation
  Platform Version                                OpenCL 1.2 
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_fp16 cl_khr_depth_images cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_icd cl_khr_image2d_from_buffer cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_intel_subgroups cl_intel_required_subgroup_size cl_intel_subgroups_short cl_khr_spir cl_intel_accelerator cl_intel_media_block_io cl_intel_driver_diagnostics cl_intel_device_side_avc_motion_estimation cl_khr_priority_hints cl_khr_throttle_hints cl_khr_create_command_queue cl_khr_fp64 cl_intel_planar_yuv cl_intel_packed_yuv cl_intel_motion_estimation cl_intel_advanced_motion_estimation cl_intel_va_api_media_sharing 
  Platform Extensions function suffix             INTEL

  Platform Name                                   Intel Gen OCL Driver
  Platform Vendor                                 Intel
  Platform Version                                OpenCL 1.2 beignet 1.3 (git-5aba95a)
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_khr_depth_images cl_khr_spir cl_khr_icd cl_intel_accelerator cl_intel_subgroups cl_intel_subgroups_short cl_khr_gl_sharing
  Platform Extensions function suffix             Intel

  Platform Name                                   Clover
  Platform Vendor                                 Mesa
  Platform Version                                OpenCL 1.1 Mesa 18.0.5
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_icd
  Platform Extensions function suffix             MESA

  Platform Name                                   Intel(R) OpenCL HD Graphics
Number of devices                                 1
  Device Name                                     Intel(R) Gen9 HD Graphics NEO
  Device Vendor                                   Intel(R) Corporation
  Device Vendor ID                                0x8086
  Device Version                                  OpenCL 1.2 NEO 
  Driver Version                                  19.14.12751
  Device OpenCL C Version                         OpenCL C 1.2 
  Device Type                                     GPU
  Device Profile                                  FULL_PROFILE
  Max compute units                               18
  Max clock frequency                             750MHz
  Device Partition                                (core)
    Max number of sub-devices                     0
    Supported partition types                     None
  Max work item dimensions                        3
  Max work item sizes                             256x256x256
  Max work group size                             256
  Preferred work group size multiple              32
  Preferred / native vector sizes                 
    char                                                16 / 16      
    short                                                8 / 8       
    int                                                  4 / 4       
    long                                                 1 / 1       
    half                                                 8 / 8        (cl_khr_fp16)
    float                                                1 / 1       
    double                                               1 / 1        (cl_khr_fp64)
  Half-precision Floating-point support           (cl_khr_fp16)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  No
  Single-precision Floating-point support         (core)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  Yes
  Double-precision Floating-point support         (cl_khr_fp64)
    Denormals                                     Yes
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 Yes
    Round to infinity                             Yes
    IEEE754-2008 fused multiply-add               Yes
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  No
  Address bits                                    32, Little-Endian
  Global memory size                              3435970560 (3.2GiB)
  Error Correction support                        No
  Max memory allocation                           1717985280 (1.6GiB)
  Unified memory for Host and Device              Yes
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       1024 bits (128 bytes)
  Global Memory cache type                        Read/Write
  Global Memory cache size                        131072
  Global Memory cache line                        64 bytes
  Image support                                   Yes
    Max number of samplers per kernel             16
    Max size for 1D images from buffer            107374080 pixels
    Max 1D or 2D image array size                 2048 images
    Base address alignment for 2D image buffers   4 bytes
    Pitch alignment for 2D image buffers          4 bytes
    Max 2D image size                             16384x16384 pixels
    Max 3D image size                             16384x16384x2048 pixels
    Max number of read image args                 128
    Max number of write image args                128
  Local memory type                               Local
  Local memory size                               65536 (64KiB)
  Max constant buffer size                        1717985280 (1.6GiB)
  Max number of constant args                     8
  Max size of kernel argument                     1024
  Queue properties                                
    Out-of-order execution                        Yes
    Profiling                                     Yes
  Prefer user sync for interop                    Yes
  Profiling timer resolution                      52ns
  Execution capabilities                          
    Run OpenCL kernels                            Yes
    Run native kernels                            No
    SPIR versions                                 1.2 
  printf() buffer size                            4194304 (4MiB)
  Built-in kernels                                block_motion_estimate_intel;block_advanced_motion_estimate_check_intel;block_advanced_motion_estimate_bidirectional_check_intel;
  Motion Estimation accelerator version	(Intel)   2
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Device Extensions                               cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_fp16 cl_khr_depth_images cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_icd cl_khr_image2d_from_buffer cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_intel_subgroups cl_intel_required_subgroup_size cl_intel_subgroups_short cl_khr_spir cl_intel_accelerator cl_intel_media_block_io cl_intel_driver_diagnostics cl_intel_device_side_avc_motion_estimation cl_khr_priority_hints cl_khr_throttle_hints cl_khr_create_command_queue cl_khr_fp64 cl_intel_planar_yuv cl_intel_packed_yuv cl_intel_motion_estimation cl_intel_advanced_motion_estimation cl_intel_va_api_media_sharing 

  Platform Name                                   Intel Gen OCL Driver
Number of devices                                 1
  Device Name                                     Intel(R) HD Graphics Broxton 0
  Device Vendor                                   Intel
  Device Vendor ID                                0x8086
  Device Version                                  OpenCL 1.2 beignet 1.3 (git-5aba95a)
  Driver Version                                  1.3
  Device OpenCL C Version                         OpenCL C 1.2 beignet 1.3 (git-5aba95a)
  Device Type                                     GPU
  Device Profile                                  FULL_PROFILE
  Max compute units                               18
  Max clock frequency                             1000MHz
  Device Partition                                (core)
    Max number of sub-devices                     1
    Supported partition types                     None, None, None
  Max work item dimensions                        3
  Max work item sizes                             512x512x512
  Max work group size                             512
  Preferred work group size multiple              16
  Preferred / native vector sizes                 
    char                                                16 / 8       
    short                                                8 / 8       
    int                                                  4 / 4       
    long                                                 2 / 2       
    half                                                 0 / 8        (cl_khr_fp16)
    float                                                4 / 4       
    double                                               0 / 2        (n/a)
  Half-precision Floating-point support           (cl_khr_fp16)
    Denormals                                     No
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 No
    Round to infinity                             No
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  No
  Single-precision Floating-point support         (core)
    Denormals                                     No
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 No
    Round to infinity                             No
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  No
  Double-precision Floating-point support         (n/a)
  Address bits                                    32, Little-Endian
  Global memory size                              4102029312 (3.82GiB)
  Error Correction support                        No
  Max memory allocation                           3076521984 (2.865GiB)
  Unified memory for Host and Device              Yes
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       1024 bits (128 bytes)
  Global Memory cache type                        Read/Write
  Global Memory cache size                        8192
  Global Memory cache line                        64 bytes
  Image support                                   Yes
    Max number of samplers per kernel             16
    Max size for 1D images from buffer            65536 pixels
    Max 1D or 2D image array size                 2048 images
    Base address alignment for 2D image buffers   4096 bytes
    Pitch alignment for 2D image buffers          1 bytes
    Max 2D image size                             8192x8192 pixels
    Max 3D image size                             8192x8192x2048 pixels
    Max number of read image args                 128
    Max number of write image args                8
  Local memory type                               Local
  Local memory size                               65536 (64KiB)
  Max constant buffer size                        134217728 (128MiB)
  Max number of constant args                     8
  Max size of kernel argument                     1024
  Queue properties                                
    Out-of-order execution                        No
    Profiling                                     Yes
  Prefer user sync for interop                    Yes
  Profiling timer resolution                      80ns
  Execution capabilities                          
    Run OpenCL kernels                            Yes
    Run native kernels                            Yes
    SPIR versions                                 1.2
  printf() buffer size                            1048576 (1024KiB)
  Built-in kernels                                __cl_copy_region_align4;__cl_copy_region_align16;__cl_cpy_region_unalign_same_offset;__cl_copy_region_unalign_dst_offset;__cl_copy_region_unalign_src_offset;__cl_copy_buffer_rect;__cl_copy_image_1d_to_1d;__cl_copy_image_2d_to_2d;__cl_copy_image_3d_to_2d;__cl_copy_image_2d_to_3d;__cl_copy_image_3d_to_3d;__cl_copy_image_2d_to_buffer;__cl_copy_image_3d_to_buffer;__cl_copy_buffer_to_image_2d;__cl_copy_buffer_to_image_3d;__cl_fill_region_unalign;__cl_fill_region_align2;__cl_fill_region_align4;__cl_fill_region_align8_2;__cl_fill_region_align8_4;__cl_fill_region_align8_8;__cl_fill_region_align8_16;__cl_fill_region_align128;__cl_fill_image_1d;__cl_fill_image_1d_array;__cl_fill_image_2d;__cl_fill_image_2d_array;__cl_fill_image_3d;
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Device Extensions                               cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_khr_depth_images cl_khr_spir cl_khr_icd cl_intel_accelerator cl_intel_subgroups cl_intel_subgroups_short cl_khr_gl_sharing cl_khr_fp16

  Platform Name                                   Clover
Number of devices                                 0

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  Intel(R) OpenCL HD Graphics
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   Success [INTEL]
  clCreateContext(NULL, ...) [default]            Success [INTEL]
  clCreateContext(NULL, ...) [other]              Success [Intel]
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  Success (1)
    Platform Name                                 Intel(R) OpenCL HD Graphics
    Device Name                                   Intel(R) Gen9 HD Graphics NEO
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  Success (1)
    Platform Name                                 Intel(R) OpenCL HD Graphics
    Device Name                                   Intel(R) Gen9 HD Graphics NEO

ICD loader properties
  ICD loader Name                                 OpenCL ICD Loader
  ICD loader Vendor                               OCL Icd free software
  ICD loader Version                              2.2.8
  ICD loader Profile                              OpenCL 1.2
	NOTE:	your OpenCL library declares to support OpenCL 1.2,
		but it seems to support up to OpenCL 2.1 too.
```

In contrast, very funny.

The device *Intel(R) Gen9 HD Graphics NEO* has all video-related extensions but no OpenGL sharing, another device *Intel(R) HD Graphics Broxton 0* has all graphics extensions but no VAAPI sharing. It means either 1 or 2, but no possible to use OpenGL to visualize the thing from 1, really sad.

**Comment by juliusmh** [comment]
Id reaally like to see the CL GL context sharing aswell :+1:  .

**Comment by dasco597** [comment]
same here, I'm really disappointed that I can't port my application to linux with intel cpus because of this. cl_khr_gl_sharing seems like such a key extension to have I'm surprised you're not supporting it.

**Comment by dasco597** [comment]
I also tried using the beignet implementation and while it does support the cl_khr_gl_sharing extension, one of the functions I required, cl_mem_new_gl_buffer was empty but for a line of code:
`FATAL ("Not implemented")`
tears all round.

**Comment by PiotrRozenfeld** [comment]
Your arguments are convincing. 
We will fit this effort into our development schedule for the remainder of this year.

**Comment by dasco597** [comment]
wow that's great to hear! your hard work is very much appreciated. best of luck to your team!

**Comment by ormf** [comment]
Hi. Yesterday my app seized working as it used cl_khr_gl_sharing with the beignet driver on linux and that seems to be disfunctional with my most recent update now. For my compositions it is absolutely crucial to work, so I'm quite devastated, as it is more or less impossible to implement my system on other Platforms/OSes (see: https://www.youtube.com/watch?v=2rWha1HTfFE&t=360s). I'd be extremely grateful if you'd implemented it and am absolutely willing to support funding if you let me know how!

**Comment by ormf** [comment]
@PiotrRozenfeld
> We will fit this effort into our development schedule for the remainder of this year.

Are there any news about whether work has been done on this and whether/when it will be implemented?

**Comment by AdamCetnerowski** [comment]
This is being actively worked on. We don't have a clear trend date, but Q1 is very likely.

**Comment by ormf** [comment]
> This is being actively worked on. We don't have a clear trend date, but Q1 is very likely.

@AdamCetnerowski: Thanks, very much appreciated!

**Comment by pioto1225** [comment]
Has there been any progress on this issue?

**Comment by smistad** [comment]
I would also very much like to see this feature implemented

**Comment by AdamCetnerowski** [comment]
Some foundational work has been done to enable the extension. At the this time the Q1 trend is no longer valid. We intend to provide a partial implementation early Q2 for evaluation. We’re looking forward to your feedback when it’s available.

**Comment by ormf** [comment]
> We’re looking forward to your feedback when it’s available.

Most definitely! Right now I'm holding back any upgrades on my system as I have to keep llvm 8.0.1 for the beignet drivers.

**Comment by AdamCetnerowski** [comment]
We have made progress towards this feature, but it was slower than initially expected. We’ll update this issue when we have something that is user-testable.”

**Comment by Apahdos** [comment]
Hi. Any news on when this will be released? Could use it for my thesis. Otherwise I need to find a workaround ^^

**Comment by PiotrRozenfeld** [comment]
While we did lay some groundwork for cl_khr_gl_sharing with MESA on Linux, we also had to increase our initial effort estimates, as  additional work to be done in MESA and Compute before we can proceed with implementation.
This does not mean we abandon the effort, but we can no longer accommodate that in the near future. 

We'll post updates when the work will resume.

**Comment by ormf** [comment]
@PiotrRozenfeld 
thanks a lot for the info! Although this is very bad news for me, I now know that I'll either have to get beignet compiled with llvm 10, or resort to installing a dual boot until this is resolved. I sincerely hope this is not the final blow to this in compute-runtime.

**Comment by Randrianasulu** [comment]
Probably not most useful comment (I'm not a developer, just user whole likes to test something new) but I was looking at AMD's ROCM sources and found there 

https://github.com/ROCm-Developer-Tools/ROCclr/blob/master/device/rocm/mesa_glinterop.h
```
/* Mesa OpenGL inter-driver interoperability interface designed for but not
 * limited to OpenCL.
```

Of course this is only smallest part of whole work .

**Comment by mathucub** [comment]
@PiotrRozenfeld any ETA of having a cl_khr_gl_sharing solution going? This extension missing causes a major problem since most of our linux machines are using Intel integrated GPUs.

**Comment by chamois94** [comment]
> While we did lay some groundwork for cl_khr_gl_sharing with MESA on Linux, we also had to increase our initial effort estimates, as additional work to be done in MESA and Compute before we can proceed with implementation.
> This does not mean we abandon the effort, but we can no longer accommodate that in the near future.
> 
> We'll post updates when the work will resume.

Any news on that ? This extension id really a must have for Manu applications.

**Comment by smlehbleh** [comment]
Also curious how this is going? We are building and deploying Intel NUC based video acquisition/processing solutions and the lack of this extension is the only reason we're still on Beignet drivers

**Comment by PiotrRozenfeld** [comment]
The work on this extension has stalled due to other work that our team is facing, unfortunately.
 
@smlehbleh - could you share a sample workload that you are looking to enable? What is your platform of interest? You are indicating this already works on Beignet - implying no additional work in MESA is needed.

**Comment by smlehbleh** [comment]
Hi Piotr,
Our use case is an Intel NUC powered cinema camera/recording device. The RAW image processing (debayering and colour transformation etc...) is performed in OpenCL kernels and the user interface is rendered with OpenGL ES. We use Beignet OpenCL drivers which support the cl_khr_gl_sharing extension to share the processed RAW image from an OpenCL buffer with OpenGL ES to draw it to the monitor screen.
We are using Ubuntu 18.04 with an Intel NUC8v7PNB and (soon) an NUC11TNBv7. The cl_khr_gl_sharing extension implementation in Beignet has been present for quite a while (Maybe a couple of years...).
For an experiment we tried the NEO CL driver and replaced our 'cl_khr_gl_sharing' GL/CL interop code with a manual additional copy from a CL buffer into a GL texture - the total CPU usage appeared to go from about 10% to 15-20% and a noticeable frame delay was introduced. The buffer was a UHD RGBA32 image (~33mb).

**Comment by smlehbleh** [comment]
Hi @PiotrRozenfeld , I was just wondering if Beignet's existing cl_khr_gl_sharing support/implementation could be ported or copied over into this driver - or if it could generally reduce the amount of work as you mentioned?

**Comment by JacekDanecki** [comment]
As I can see in Beignet documentation cl_khr_gl_sharing is partially supported, and allow to create memory objects from OpenGL buffers or 2D textures.
If I understand your use case correctly you're using cl/gl sharing differently, and you share OpenCL buffers with OpenGL, so memory for buffers is allocated by OpenCL not by OpenGL. Could you clarify?

**Comment by smlehbleh** [comment]
Hi Jacek,
We are using the Beignet implementation of cl_khr_gl_sharing in the standard intended use case - we create an OpenGL texture and then create an OpenCL buffer from it using 'clCreateFromGLTexture'.
I realise in the previous comment it seems like we're directly sharing an OpenCL buffer with OpenGL, but our currently active implementation is an OpenCL buffer created from an OpenGL texture (clCreateFromGLTexture). 
This allows us to draw the GL texture to the screen with OpenGL after an OpenCL kernel writes to the underlying buffer without any additional 'copy'.

**Comment by ormf** [comment]
Hi Jacek,

Am Mittwoch, den 01. September 2021 um 05:26:00 Uhr (-0700) schrieb
Jacek Danecki:
> As I can see in Beignet documentation cl_khr_gl_sharing is partially
> supported, and allow to create memory objects from OpenGL buffers or
> 2D textures.
>
> If I understand your use case correctly you're using cl/gl sharing
> differently, and you share OpenCL buffers with OpenGL, so memory for
> buffers is allocated by OpenCL not by OpenGL. Could you clarify?

 it's unclear whether you address me in your mail.

The way I used it with the Beignet driver was to first allocate the
buffers in OpenGL (using "gen-buffer") and then "recreating" them in
OpenCL (within the opengl context with the allocated OpenGL buffer
mapped) by calling the OpenCL routine "create-buffer" with the
USE-HOST-PTR flag.

Is that what you were asking?

Best,
Orm

**Comment by mathucub** [comment]
I'm going to chime in, even though it has taken so long to get any traction that I've change jobs so no longer care.
GL and CL are supposed to be able to interop with calls like clCreateFromGLTexture and stay in the GPU address space.

This allows for work to be done in CL, but presented in GL in realtime -- which any copy to host techniques will quash. This has been as baseline feature in nVidia and AMD's implementation for years.

The Intel stack on Linux does not match what the Intel stack on Windows can do, that is the baseline issue of what is wrong. 
They have to be in parity otherwise software is trapped on Windows since it can't be ported.

**Comment by JacekDanecki** [comment]
In Beignet code there is example: https://github.com/intel/beignet/blob/master/examples/gl_buffer_sharing/gl_buffer_sharing.cpp, and it looks similar to your usages. After small changes in Beignet code I was able to run this example on Ubuntu 18.04 and it worked correctly. Looking at sharing implementation in Beignet I can see it uses https://www.khronos.org/registry/EGL/extensions/MESA/EGL_MESA_image_dma_buf_export.txt, so I suppose it should be enough on mesa side to implement basic sharing in Neo.

**Comment by JacekDanecki** [comment]
I've created new repository [cl-gl-tests](https://github.com/JacekDanecki/cl-gl-tests) where I've pushed modified example from Beignet. I'll use this repo to test cl/gl implementation in Neo, so if you see any potential issues with this test in comparison to your usage scenario, please let me know. In first step I'll prepare some changes in Neo to run only tests from this repo.
When I've any working Neo code I'll push it to my fork: https://github.com/JacekDanecki/compute-runtime on clgl branch.

**Comment by JacekDanecki** [comment]
With commit https://github.com/JacekDanecki/compute-runtime/commit/b95bc31fd7547cf5da7560bea822dca159c9e07d test gl_buffer_sharing from: https://github.com/JacekDanecki/cl-gl-tests is working now. 
This is still very experimental code, under development, there are some hard coded values, and some parts of code are stubbed. I've only executed gl_buffer_sharing test so far.

If you have any tests, you can share, I can focus on enabling them.

**Comment by JacekDanecki** [comment]
@smlehbleh @ormf any feedback on experimental code/test I've provided?

**Comment by smlehbleh** [comment]
Hi @JacekDanecki, sorry was pulled onto other things, but I'm back on this now. I'm not too familiar with building the necessary .deb packages from the compute-runtime repo. Are you able to enable CI on your forked repo to make some .deb packages show up in the releases pane? Then I'll give them a go with some gl_cl interop samples I've got.
Cheers, Russell

**Comment by smlehbleh** [comment]
Ubuntu 20.04.3 on an 11th Gen Intel NUC

**Comment by JacekDanecki** [comment]
I'll rebase my current code to the latest Neo release and push to my fork, so you can use igc and gmmlib packages from release.

**Comment by smlehbleh** [comment]
Sounds great, thanks, let me know when to grab the packages

**Comment by Abdob** [comment]
"Right now, Intel does not support surface sharing on Linux* for example. Customer requests can drive changes to this decision."
Last Updated: 12/15/2014

By Adam T Lake, Robert M Ioffe

https://www.intel.com/content/www/us/en/developer/articles/technical/sharing-surfaces-between-opencl-and-opengl-43-on-intel-processor-graphics-using-implicit.html

Is this still the case today?

**Comment by smlehbleh** [comment]
@Abdob, this is my understanding/experience:

It depends on the driver. Beignet OCL driver (1.3.2-6) supports the gl/cl sharing extension. The NEO Compute runtime OCL driver does not. However, @JacekDanecki has an experimental/wip branch with the feature using I think an approach similar to the Beignet driver. 

Unfortunately the Beignet driver doesn't appear to support 11th Gen Intel GPUs (Xe) so for the latest hardware, cl/gl sharing support is not in any official driver release.

**Comment by Abdob** [comment]
Hi @smlehbleh 

My intel GPU is:
glxinfo | grep Mesa
client glx vendor string: Mesa Project and SGI
    Device: Mesa Intel(R) UHD Graphics 630 (CML GT2) (0x9bc5)
OpenGL renderer string: Mesa Intel(R) UHD Graphics 630 (CML GT2)
OpenGL core profile version string: 4.6 (Core Profile) Mesa 21.0.3
OpenGL version string: 4.6 (Compatibility Profile) Mesa 21.0.3
OpenGL ES profile version string: OpenGL ES 3.2 Mesa 21.0.3

I install beignet-opencl-icd and clinfo and I get this:

wave@wave:~$ clinfo
Number of platforms                               1
  Platform Name                                   Intel Gen OCL Driver
  Platform Vendor                                 Intel
  Platform Version                                OpenCL 2.0 beignet 1.3
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_khr_depth_images cl_khr_spir cl_khr_icd cl_intel_accelerator cl_intel_subgroups cl_intel_subgroups_short
  Platform Extensions function suffix             Intel
beignet-opencl-icd: no supported GPU found, this is probably the wrong opencl-icd package for this hardware
(If you have multiple ICDs installed and OpenCL works, you can ignore this message)

  Platform Name                                   Intel Gen OCL Driver
Number of devices                                 0

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  No platform
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   No platform
  clCreateContext(NULL, ...) [default]            No platform
  clCreateContext(NULL, ...) [other]              No platform
beignet-opencl-icd: no supported GPU found, this is probably the wrong opencl-icd package for this hardware
(If you have multiple ICDs installed and OpenCL works, you can ignore this message)
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_DEFAULT)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  No devices found in platform
beignet-opencl-icd: no supported GPU found, this is probably the wrong opencl-icd package for this hardware
(If you have multiple ICDs installed and OpenCL works, you can ignore this message)
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  No devices found in platform
beignet-opencl-icd: no supported GPU found, this is probably the wrong opencl-icd package for this hardware
(If you have multiple ICDs installed and OpenCL works, you can ignore this message)
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  No devices found in platform

Is cl_khr_gl_sharing suppose to be listed as a device extension?

Thanks,
Abdo

**Comment by JacekDanecki** [comment]
@Abdob  cl_khr_gl_sharing is listed on my setup with Beignet
In case of CML support in Beignet see: https://bugs.launchpad.net/ubuntu/+source/beignet/+bug/1905340

glxinfo grep Mesa
```
client glx vendor string: Mesa Project and SGI
    Device: Mesa Intel(R) Iris(R) Pro Graphics 580 (SKL GT4) (0x193b)
OpenGL renderer string: Mesa Intel(R) Iris(R) Pro Graphics 580 (SKL GT4)
OpenGL core profile version string: 4.6 (Core Profile) Mesa 21.0.3
OpenGL version string: 4.6 (Compatibility Profile) Mesa 21.0.3
OpenGL ES profile version string: OpenGL ES 3.2 Mesa 21.0.3
```
clinfo | grep cl_khr_gl
```
  Platform Extensions                             cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_khr_depth_images cl_khr_spir cl_khr_icd cl_intel_accelerator cl_intel_subgroups cl_intel_subgroups_short cl_intel_required_subgroup_size cl_intel_media_block_io cl_intel_planar_yuv cl_khr_gl_sharing
  Device Extensions                               cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_khr_depth_images cl_khr_spir cl_khr_icd cl_intel_accelerator cl_intel_subgroups cl_intel_subgroups_short cl_intel_required_subgroup_size cl_intel_media_block_io cl_intel_planar_yuv cl_khr_gl_sharing cl_khr_fp16 cl_intel_device_side_avc_motion_estimation
```
clinfo
```
  Platform Extensions                             cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_khr_depth_images cl_khr_spir cl_khr_icd cl_intel_accelerator cl_intel_subgroups cl_intel_subgroups_short cl_intel_required_subgroup_size cl_intel_media_block_io cl_intel_planar_yuv cl_khr_gl_sharing
  Device Extensions                               cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_khr_depth_images cl_khr_spir cl_khr_icd cl_intel_accelerator cl_intel_subgroups cl_intel_subgroups_short cl_intel_required_subgroup_size cl_intel_media_block_io cl_intel_planar_yuv cl_khr_gl_sharing cl_khr_fp16 cl_intel_device_side_avc_motion_estimation
root@gubuntuPC:/home/jdanecki# cat clinfo.txt
Number of platforms                               1
  Platform Name                                   Intel Gen OCL Driver
  Platform Vendor                                 Intel
  Platform Version                                OpenCL 2.0 beignet 1.4 (git-419c0417)
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_khr_depth_images cl_khr_spir cl_khr_icd cl_intel_accelerator cl_intel_subgroups cl_intel_subgroups_short cl_intel_required_subgroup_size cl_intel_media_block_io cl_intel_planar_yuv cl_khr_gl_sharing
  Platform Extensions function suffix             Intel

  Platform Name                                   Intel Gen OCL Driver
Number of devices                                 1
  Device Name                                     Intel(R) HD Graphics Skylake Halo GT4
  Device Vendor                                   Intel
  Device Vendor ID                                0x8086
  Device Version                                  OpenCL 2.0 beignet 1.4 (git-419c0417)
  Driver Version                                  1.4
  Device OpenCL C Version                         OpenCL C 2.0 beignet 1.4 (git-419c0417)
  Device Type                                     GPU
  Device Profile                                  FULL_PROFILE
  Device Available                                Yes
  Compiler Available                              Yes
  Linker Available                                Yes
  Max compute units                               72
  Max clock frequency                             1000MHz
  Device Partition                                (core)
    Max number of sub-devices                     1
    Supported partition types                     None, None, None
    Supported affinity domains                    (n/a)
  Max work item dimensions                        3
  Max work item sizes                             512x512x512
  Max work group size                             256
  Preferred work group size multiple              16
  Sub-group sizes (Intel)                         8, 16
  Preferred / native vector sizes                 
    char                                                16 / 8       
    short                                                8 / 8       
    int                                                  4 / 4       
    long                                                 2 / 2       
    half                                                 0 / 8        (cl_khr_fp16)
    float                                                4 / 4       
    double                                               0 / 2        (n/a)
  Half-precision Floating-point support           (cl_khr_fp16)
    Denormals                                     No
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 No
    Round to infinity                             No
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
  Single-precision Floating-point support         (core)
    Denormals                                     No
    Infinity and NANs                             Yes
    Round to nearest                              Yes
    Round to zero                                 No
    Round to infinity                             No
    IEEE754-2008 fused multiply-add               No
    Support is emulated in software               No
    Correctly-rounded divide and sqrt operations  No
  Double-precision Floating-point support         (n/a)
  Address bits                                    32, Little-Endian
  Global memory size                              4294967296 (4GiB)
  Error Correction support                        No
  Max memory allocation                           3221225472 (3GiB)
  Unified memory for Host and Device              Yes
  Shared Virtual Memory (SVM) capabilities        (core)
    Coarse-grained buffer sharing                 Yes
    Fine-grained buffer sharing                   No
    Fine-grained system sharing                   No
    Atomics                                       No
  Minimum alignment for any data type             128 bytes
  Alignment of base address                       1024 bits (128 bytes)
  Preferred alignment for atomics                 
    SVM                                           0 bytes
    Global                                        0 bytes
    Local                                         0 bytes
  Max size for global variable                    65536 (64KiB)
  Preferred total size of global vars             65536 (64KiB)
  Global Memory cache type                        Read/Write
  Global Memory cache size                        8192 (8KiB)
  Global Memory cache line size                   64 bytes
  Image support                                   Yes
    Max number of samplers per kernel             16
    Max size for 1D images from buffer            65536 pixels
    Max 1D or 2D image array size                 2048 images
    Base address alignment for 2D image buffers   4096 bytes
    Pitch alignment for 2D image buffers          1 pixels
    Max 2D image size                             8192x8192 pixels
    Max planar YUV image size                     8192x8192 pixels
    Max 3D image size                             8192x8192x2048 pixels
    Max number of read image args                 128
    Max number of write image args                8
    Max number of read/write image args           8
  Max number of pipe args                         16
  Max active pipe reservations                    1
  Max pipe packet size                            1024
  Local memory type                               Local
  Local memory size                               65536 (64KiB)
  Max number of constant args                     8
  Max constant buffer size                        134217728 (128MiB)
  Max size of kernel argument                     1024
  Queue properties (on host)                      
    Out-of-order execution                        No
    Profiling                                     Yes
  Queue properties (on device)                    
    Out-of-order execution                        Yes
    Profiling                                     Yes
    Preferred size                                16384 (16KiB)
    Max size                                      262144 (256KiB)
  Max queues on device                            1
  Max events on device                            1024
  Prefer user sync for interop                    Yes
  Profiling timer resolution                      80ns
  Execution capabilities                          
    Run OpenCL kernels                            Yes
    Run native kernels                            Yes
    SPIR versions                                 1.2
  printf() buffer size                            1048576 (1024KiB)
  Built-in kernels                                __cl_copy_region_align4;__cl_copy_region_align16;__cl_copy_region_unalign_same_offset;__cl_copy_region_unalign_dst_offset;__cl_copy_region_unalign_src_offset;__cl_copy_buffer_rect;__cl_copy_buffer_rect_align4;__cl_copy_image_1d_to_1d;__cl_copy_image_2d_to_2d;__cl_copy_image_3d_to_2d;__cl_copy_image_2d_to_3d;__cl_copy_image_3d_to_3d;__cl_copy_image_2d_to_buffer;__cl_copy_image_2d_to_buffer_align4;__cl_copy_image_2d_to_buffer_align16;__cl_copy_image_3d_to_buffer;__cl_copy_image_3d_to_buffer_align4;__cl_copy_image_3d_to_buffer_align16;__cl_copy_buffer_to_image_2d;__cl_copy_buffer_to_image_2d_align4;__cl_copy_buffer_to_image_2d_align16;__cl_copy_buffer_to_image_3d;__cl_copy_buffer_to_image_3d_align4;__cl_copy_buffer_to_image_3d_align16;__cl_copy_image_1d_array_to_1d_array;__cl_copy_image_2d_array_to_2d_array;__cl_copy_image_2d_array_to_2d;__cl_copy_image_2d_array_to_3d;__cl_copy_image_2d_to_2d_array;__cl_copy_image_3d_to_2d_array;__cl_fill_region_unalign;__cl_fill_region_align2;__cl_fill_region_align4;__cl_fill_region_align8_2;__cl_fill_region_align8_4;__cl_fill_region_align8_8;__cl_fill_region_align8_16;__cl_fill_region_align128;__cl_fill_image_1d;__cl_fill_image_1d_array;__cl_fill_image_2d;__cl_fill_image_2d_array;__cl_fill_image_3d;
    Device-side AVC Motion Estimation version     <printDeviceInfo:165: get CL_DEVICE_AVC_ME_VERSION_INTEL : error -30>
      Supports texture sampler use                <printDeviceInfo:166: get CL_DEVICE_AVC_ME_SUPPORTS_TEXTURE_SAMPLER_USE_INTEL : error -30>
      Supports preemption                         <printDeviceInfo:167: get CL_DEVICE_AVC_ME_SUPPORTS_PREEMPTION_INTEL : error -30>
  Device Extensions                               cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_khr_depth_images cl_khr_spir cl_khr_icd cl_intel_accelerator cl_intel_subgroups cl_intel_subgroups_short cl_intel_required_subgroup_size cl_intel_media_block_io cl_intel_planar_yuv cl_khr_gl_sharing cl_khr_fp16 cl_intel_device_side_avc_motion_estimation

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  Intel Gen OCL Driver
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   Success [Intel]
  clCreateContext(NULL, ...) [default]            Success [Intel]
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_DEFAULT)  Success (1)
    Platform Name                                 Intel Gen OCL Driver
    Device Name                                   Intel(R) HD Graphics Skylake Halo GT4
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  Success (1)
    Platform Name                                 Intel Gen OCL Driver
    Device Name                                   Intel(R) HD Graphics Skylake Halo GT4
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  Success (1)
    Platform Name                                 Intel Gen OCL Driver
    Device Name                                   Intel(R) HD Graphics Skylake Halo GT4

ICD loader properties
  ICD loader Name                                 OpenCL ICD Loader
  ICD loader Vendor                               OCL Icd free software
  ICD loader Version                              2.2.11
  ICD loader Profile                              OpenCL 2.1
```

**Comment by Abdob** [comment]
Hi @JacekDanecki 

Thank you for the follow up. 

According to the link you provided beignet 1.3.2-8 fixed the issue. I installed 1.4 from source and it still doesn't show. We decided to move forward with another API and will no longer attempt to get the OpenCL interoperability to work.

wave@wave:~/Documents/ocl/beignet/build$ clinfo
Number of platforms                               1
  Platform Name                                   Intel Gen OCL Driver
  Platform Vendor                                 Intel
  Platform Version                                OpenCL 2.0 beignet 1.4 (git-419c0417)
  Platform Profile                                FULL_PROFILE
  Platform Extensions                             cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_3d_image_writes cl_khr_image2d_from_buffer cl_khr_depth_images cl_khr_spir cl_khr_icd cl_intel_accelerator cl_intel_subgroups cl_intel_subgroups_short cl_intel_required_subgroup_size cl_intel_media_block_io cl_intel_planar_yuv
  Platform Extensions function suffix             Intel
cl_get_gt_device(): error, unknown device: 9bc5

  Platform Name                                   Intel Gen OCL Driver
Number of devices                                 0

NULL platform behavior
  clGetPlatformInfo(NULL, CL_PLATFORM_NAME, ...)  No platform
  clGetDeviceIDs(NULL, CL_DEVICE_TYPE_ALL, ...)   No platform
  clCreateContext(NULL, ...) [default]            No platform
  clCreateContext(NULL, ...) [other]              No platform
cl_get_gt_device(): error, unknown device: 9bc5
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_DEFAULT)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CPU)  No devices found in platform
cl_get_gt_device(): error, unknown device: 9bc5
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_GPU)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ACCELERATOR)  No devices found in platform
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_CUSTOM)  No devices found in platform
cl_get_gt_device(): error, unknown device: 9bc5
  clCreateContextFromType(NULL, CL_DEVICE_TYPE_ALL)  No devices found in platform

**Comment by eero-t** [comment]
> According to the link you provided beignet 1.3.2-8 fixed the issue.

@Abdob That link was to Ubuntu bug tracker, and that fix was into a _Ubuntu_ (and Debian) 1.3.2-8 package of beignet, not to beignet upstream project.

> I installed 1.4 from source and it still doesn't show.
...
> Platform Version OpenCL 2.0 beignet 1.4 (git-419c0417)
> cl_get_gt_device(): error, unknown device: 9bc5

If you built it from upstream, and it's this one (which I found by googling), quite obviously it's missing CML support:
https://github.com/intel/beignet/commits/master/src/cl_device_data.h

As the PR for that has not been merged: https://github.com/intel/beignet/pull/20

You should have built e.g. the Debian version (with Debian tools, so that CML support patch gets applied on top of upstream sources): https://salsa.debian.org/opencl-team/beignet

**Comment by smlehbleh** [comment]
Hi  @JacekDanecki, I was wondering if you were able to build and generate the NEO compute runtime .deb files for Ubuntu 20.04from your fork with the experimental/wip cl/gl sharing?

**Comment by JacekDanecki** [comment]
@smlehbleh I've pushed current experimental implementation rebased on release https://github.com/intel/compute-runtime/releases/tag/22.06.22433 to clgl-fork branch in repository: https://github.com/JacekDanecki/compute-runtime/tree/clgl-fork
You can  use gmmlib and igc packages from Neo release.

Neo can be compiled with commands:
```
mkdir build
cd build
cmake -DCMAKE_BUILD_TYPE=Debug -DDISABLE_WDDM_LINUX=1 -DBUILD_WITH_L0=1 -DNEO_DISABLE_LD_GOLD=1 ..
make  igdrcl_dll
```
Once compiled you can create file neo.icd in /etc/OpenCL/vendors directory containing full path to libigdrcl.so library.

**Comment by JacekDanecki** [comment]
I'm leaving compute team, so will not work on cl/gl sharing anymore.

**Comment by Abdob** [comment]
@smlehbleh 

I'm able to build  https://salsa.debian.org/opencl-team/beignet with an ubuntu 16 docker container, I see that it has the cl_khr_gl_sharing extension.

Thanks,
Abdo

**Comment by AwesomeCronk** [comment]
Tried Beignet, it doesn't support my CPU. -_-
Also tried building @JacekDanecki 's fork, but even though I installed IGC in order to run normal NEO (which works), the CMake script failed to find it.

**Comment by smlehbleh** [comment]
Hi @JacekDanecki - is it possible someone else will pick up the work? It seems like not a massive amount of dev work is needed to get this highly desirable feature over the line.

**Comment by JacekDanecki** [comment]
@smlehbleh Please wait for an answer from the team, as I'm not a member of compute team anymore, so can't answer your question.
@AwesomeCronk It'll be helpful if you provide log from cmake.

**Comment by AwesomeCronk** [comment]
I'll try and get one this afternoon @JacekDanecki . Should I put it here?

**Comment by JacekDanecki** [comment]
@AwesomeCronk copy it here, or add the link to the log.

**Comment by AwesomeCronk** [comment]
@JacekDanecki I believe this is the right log. (`build/CMakeFiles/CMakeOutput.log`)
[CMakeOutput.log](https://github.com/intel/compute-runtime/files/8530732/CMakeOutput.log)

**Comment by PoggPoggBeanMan** [comment]
Would love to see this as well.

**Comment by seadra** [comment]
The lack of this extension has long prevented DaVinci Resolve and DaVinci Resolve Studio from functioning on Linux (the editor starts, but no visuals from the videos appear).

**Comment by kallaballa** [comment]
> @smlehbleh I've pushed current experimental implementation rebased on release https://github.com/intel/compute-runtime/releases/tag/22.06.22433 to clgl-fork branch in repository: https://github.com/JacekDanecki/compute-runtime/tree/clgl-fork You can use gmmlib and igc packages from Neo release.
> 
> Neo can be compiled with commands:
> 
> ```
> mkdir build
> cd build
> cmake -DCMAKE_BUILD_TYPE=Debug -DDISABLE_WDDM_LINUX=1 -DBUILD_WITH_L0=1 -DNEO_DISABLE_LD_GOLD=1 ..
> make  igdrcl_dll
> ```
> 
> Once compiled you can create file neo.icd in /etc/OpenCL/vendors directory containing full path to libigdrcl.so library.

Based on your work I've got a full example of OpenGL -> OpenCL -> VAAPI working on my Tigerlake (i7-1160G7) CPU . Stay tuned for full instructions and an example based on OpenCV.

https://github.com/kallaballa/compute-runtime

**Comment by kallaballa** [comment]
I had to fork the 4.x branch of OpenCV and still work on cleaning my fork up, but once I pushed that (and the instructions) you'll be able to do something like the following on a Tigerlake CPU (and maybe others) running Linux:

```c++
int main(int argc, char **argv) {
	using namespace kb;

	if (!va::open_display())
		throw std::runtime_error("Failed to open VA display for CL-VA interoperability");

	va::check_if_YUV420_available();

	cv::va_intel::ocl::initializeContextFromVA(va::display, true);
	cv::ocl::OpenCLExecutionContext vaContext = cv::ocl::OpenCLExecutionContext::getCurrentRef().cloneWithNewQueue();

	VideoWriter video("out.mkv", CAP_FFMPEG, VideoWriter::fourcc('V', 'P', '9', '0'), FPS, Size(WIDTH, HEIGHT), {
			VIDEOWRITER_PROP_HW_DEVICE, 0,
			VIDEOWRITER_PROP_HW_ACCELERATION, VIDEO_ACCELERATION_VAAPI,
			VIDEOWRITER_PROP_HW_ACCELERATION_USE_OPENCL, 1
	});

	egl::init_egl();
	gl::init_gl();
	cv::ocl::OpenCLExecutionContext glContext = cv::ocl::OpenCLExecutionContext::getCurrentRef().cloneWithNewQueue();

	UMat frameBuffer(HEIGHT, WIDTH, CV_8UC4, Scalar::all(0));
	UMat grey(HEIGHT, WIDTH, CV_8UC1, Scalar::all(0));
	UMat bgr(HEIGHT, WIDTH, CV_8UC3, Scalar::all(0));

	while (true) {
		//Draw a large triangle on blue background using OpenGL
		glContext.bind();
		gl::begin_gl();
		glClearColor(1.0f, 0.0f, 0.0f, 1.0f);
		glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
		glBegin(GL_POLYGON);
		glColor3f(1, 0, 0);
		glVertex3f(-0.6, -0.75, 0.5);
		glColor3f(0, 1, 0);
		glVertex3f(0.6, -0.75, 0);
		glColor3f(0, 0, 1);
		glVertex3f(0, 0.75, 0);
		glEnd();
		glFlush();
		gl::end_gl();
		gl::fetch_frame_buffer(frameBuffer); //hand over the data (GPU 2 GPU) to OpenCV/OpenCL

		//do colorspace conversions using OpenCL in the background
		cvtColor(frameBuffer, grey, COLOR_BGRA2GRAY);
		cvtColor(grey, bgr, COLOR_BGRA2BGR);

		vaContext.bind();
		//encode the frame using VAAPI on the GPU.
		video.write(bgr);
	}

	return 0;
}
```

**Comment by kallaballa** [comment]
this branch contains the changes it required to get OpenCV to play nicely: https://github.com/kallaballa/opencv/tree/GCV
this branch contains the changes to compute-runtime it needed: https://github.com/kallaballa/compute-runtime/tree/clgl-fork
and this repo contains the full demo source code: https://github.com/kallaballa/GCV

Proper instructions follow on the weekend :)

**Comment by kallaballa** [comment]
I was able to merge with upstream/22.42.24548. That means we have everything everybody has, with additional linux CL/GL interop support.

**Comment by kallaballa** [comment]
Now I am working on creating packages for easy installing

**Comment by kallaballa** [comment]
I have made a binary release (https://github.com/kallaballa/compute-runtime/releases/tag/22.42.24548-clgl) that is on par with the latest official release. Please test and give me feedback!
Instructions are in the readme: https://github.com/kallaballa/compute-runtime/#readme

**Comment by kquote03** [comment]
Hello, is there any update on this issue ?
Thank you

**Comment by JablonskiMateusz** [comment]
Hi folks
I prepared PR (#611) to apply @kallaballa  code to the top of neo repo. Please review the PR and leave some feedback.

**Comment by m911t** [comment]
hey, hopefully this can be supported officially soon! The intel platform is great and getting better by the minute!

**Comment by kallaballa** [comment]
FYI, the feature is merged. I don't know when it will be in the distribution repositories.

**Comment by allinurl** [comment]
@kallaballa Thanks a lot for your work on this! Do you know what version shipped this?

**Comment by kallaballa** [comment]
The feature was introduced with this commit: https://github.com/intel/compute-runtime/commit/4919c7c9fba1a2fda4b9489ceb733b4917840638
I don't believe it is included in a release yet.

**Comment by L911P** [comment]
not yet, hopefully next one!

**Comment by RannyBergamotte** [comment]
Yay! Will Davinci Resolve finally be able to run on intel integrated graphics with this?

**Comment by kallaballa** [comment]
The comment in the table of [this section](https://wiki.archlinux.org/title/DaVinci_Resolve#Installation) of the arch wiki suggests so.

"Launches normally, can open a project. But unable to start playing the timeline (even without video tracks) in the Cut and in the Edit pages. However in the Fairlight page it is able to play timeline.
Tested with Intel Core i7-8700 CPU. This is due to the 'cl_khr_gl_sharing' extension not being implemented."

**Comment by smunaut** [comment]
Is there any other dependencies that need to be updated for this to work ?

I build the latest release (which has this patch) and I see the extension now being reported, but it doesn't work. Context creation fails and I'm not sure why.

Or is there a small example of it that known to work so I can see what I'm doing wrong ?

**Comment by kallaballa** [comment]
Are you using EGL? At the moment only EGL is supported.
Also:
* Only GL_TEXTURE_2D target
*Only GL internal_formats: GL_RGBA, GL_RGBA8, GL_RGBA16F, GL_RGB

**Comment by smunaut** [comment]
Arf :sob: 

EGL I could easily switch so the context creation works, but the rest, not so much.

I need more like GL_R32F .
And also in addition to GL_TEXTURE2D I use  clCreateFromGLBuffer to share a GL_ARRAY_BUFFER, so I'm assuming this isn't going to work.

Damn, I was so happy when I saw this but I guess more waiting needed. I hope it will be extended to be more complete.

BTW I think that also means Davinci Resolve won't work, it uses GLX AFAICT.

**Comment by smunaut** [comment]
So indeed with EGL I can create the context with sharing enabled just fine. Didnd't actually try anything more since for my application I'd need more formats.

Any reason for the image format limitation ?
clGetSupportedImageFormats returns plenty of other formats and the GlSharing::glToCLFormats also has many more formats defined. So I'm curious where the limitation comes from ?  
I mean I see the `switch` in `createSharedGlTexture` filtering the formats, but why ? I traced the code path down to `SurfaceFormats::surfaceFormats` and a bunch of other formats are in there AFAICT.


I also checked Davinci Resolve and no-joy, doesn't work. Probably needs GLX (looking at linked libs, plenty of GLX reference and not any EGL references).

**Comment by kallaballa** [comment]
Give it a try and just un-filter the formats? At the moment I don't have time :)

**Comment by kallaballa** [comment]
> not yet, hopefully next one!

It is out: https://github.com/intel/compute-runtime/releases/tag/23.13.26032.30

:)

**Comment by pearsche** [comment]
that's nice but, not usable with Davinci Resolve yet, sadly

**Comment by jeckathi** [comment]
How long will it take for Davinci Resolve to work?

**Comment by smunaut** [comment]
Until someone works on it ... AFAIK no one is at the moment.
At the _very_ least support for it on GLX (and not only EGL) would be required.

**Comment by SimonBrandner** [comment]
I am somewhat confused/lost - where is the GLX feature missing, is there a feature request open for it?

**Comment by kallaballa** [comment]
Not yet. And at the moment I don't have time to implement it. Would you open a feature request?

**Comment by kallaballa** [comment]
Btw. current proprietary nvidia drivers (version: 530.41.03) support cl_khr_gl_sharing but only with GLX. It's a transitional time for GLX/EGL and it's gonna take time till it settles and till i find time to dive back in :)

**Comment by smunaut** [comment]
Is there some demo code for this feature ?  All my attempts to share a buffer resulted in the software just hanging when trying to create the buffer from the GL object.

**Comment by kallaballa** [comment]
I use it extensively in conjunction with opencv, but EGL based CL-GL support (https://github.com/opencv/opencv/pull/22704) is not merged yet. So for their interop example (https://fossies.org/linux/opencv/samples/opengl/opengl_interop.cpp) to work you need to use my fork (https://github.com/kallaballa/opencv/tree/GCV) and of course replace GLX calls in the example with EGL calls.

If you want step by step instructions you could go the full way and build my (still) inofficial opencv module (https://viel-zu.org/opencv/doxygen/html/d7/dfc/v4d.html) which uses cl-gl successfully.

There is also boost interoperability (https://www.boost.org/doc/libs/1_62_0/libs/compute/doc/html/boost_compute/interop.html) which you could check out for examples.

**Comment by kallaballa** [comment]
Maybe you want to post some code? I'd happily try to help you.

**Comment by smunaut** [comment]
Thanks. I'll try to write a minimal example tomorrow.

**Comment by smunaut** [comment]
Ok, so I managed to get it to work (I had the wrong internal format previously).

However, I still don't get why only GL_RGBA, GL_RGBA8, GL_RGBA16F, GL_RGB are working.
I removed the check to try others. I also followed the code path and I don't get why others are not supported, AFAICT everywhere where the internal format is used, there is provision for the other internal format and yet if you try to use GL_RGBA32F for instance, it just hangs :(

**Comment by kallaballa** [comment]
Could you maybe attach at gdb backtrace or something similar? That way I could jump right in and try to understand what is going on.

**Comment by smunaut** [comment]
Of course

```
(gdb) bt
#0  0x00007f243ae1ab1b in __lll_lock_wait () from /lib64/libc.so.6
#1  0x00007f243ae20e12 in pthread_mutex_lock () from /lib64/libc.so.6
#2  0x00007f243a5de1f9 in mtx_lock () from /usr/lib64/libEGL_mesa.so.0
#3  0x00007f243a5cc92f in dri2_destroy_image_khr () from /usr/lib64/libEGL_mesa.so.0
#4  0x00007f243a5bd962 in _eglDestroyImageCommon () from /usr/lib64/libEGL_mesa.so.0
#5  0x00007f2405d4faf0 in NEO::GlTexture::createSharedGlTexture(NEO::Context*, unsigned long, unsigned int, int, unsigned int, int*) () from /usr/lib64/intel-opencl/libigdrcl.so
#6  0x00007f2405d4a6a7 in clCreateFromGLTexture () from /usr/lib64/intel-opencl/libigdrcl.so
#7  0x0000560e0e97f733 in cl_do_init () at main.c:217
#8  0x0000560e0e97f353 in main (argc=<optimized out>, argv=<optimized out>) at main.c:255
```

**Comment by smunaut** [comment]
I dug a bit into that and noticed two things : 
* The hang I notice is probably some bug in the error path. _eglDestroyImageCommon is only called if eglExportDMABUFImageMESA failed. So it should definitely not hang and it should return the error code instead of hanging but it's also not the root of my problem, ideally that call should work.
* eglExportDMABUFImageMESA fails because dri2_can_export_dma_buf_image returns FALSE. And that's the root of the issue.
* And the above eventually falls to missing formats in dri2_format_table ...  but I have no idea if I can just add format to that table or what it implies or whatever ...

**Comment by smunaut** [comment]
(Moved discussion to a new issue since ...)
